sdl_viewer/src/graphic/moving_window_texture.rs

```diff
@@ -0,0 +1,373 @@
+use crate::opengl;
+use crate::opengl::types::{GLint, GLuint};
+use cgmath::Vector2;
+use image::{GenericImageView, ImageBuffer, LumaA, Pixel, Rgba, SubImage};
+use num_integer::Integer;
+use std::convert::{TryFrom, TryInto};
+use std::ffi::{c_void, CString};
+use std::marker::PhantomData;
+use std::rc::Rc;
+
+/// A square texture that functions like a moving window into a larger texture.
+/// Incrementality is achieved by using wraparound indexing and overwriting the
+/// parts that moved off the texture with the newly loaded parts.
+/// See `https://developer.nvidia.com/gpugems/GPUGems2/gpugems2_chapter02.html`,
+/// section 2.4 for an illustration, but note that this page does not describe
+/// the present implementation, which is simpler.
+/// Also publishes an `<texture_name>_texture_offset` uniform whose unit is
+/// pixels (not UV coordinates).
+///
+/// I'm told that a more efficient alternative would be to use a PBO (pixel
+/// buffer object), so a future version could use that instead.
+pub struct GlMovingWindowTexture<P: Pixel> {
+    id: GLuint,
+    gl: Rc<opengl::Gl>,
+    size: i32,
+    u_texture_offset: GlUniform<Vector2<i32>>,
+    pixel_type: PhantomData<P>,
+}
+
+/// Specifies the offsets in the destination, as well as the image to be pasted
+/// there.
+struct UpdateRegion<'a, P: Pixel + 'static> {
+    x: i32,
+    y: i32,
+    pixels: SubImage<&'a ImageBuffer<P, Vec<P::Subpixel>>>,
+}
+
+impl<'a, P: Pixel + 'static> UpdateRegion<'a, P> {
+    /// Splits an image with offset into update regions. See
+    /// [`incremental_update()`] for a detailed explanation.
+    /// This works not only for full-width/full-height strips, but
+    /// here are vertical (left) and horizontal (right) strips
+    /// visualized:
+    ///
+    /// ```text              
+    ///     +---+-+          
+    ///     | 3 |4|              
+    ///     +---+-+              +---+---------+
+    ///     |   | |              | 3 |    4    |
+    ///     | 1 |2|              +---+---------+
+    ///     |   | |              | 1 |    2    |
+    ///     +---+-+              +---+---------+
+    ///        |                        |
+    ///        V                        V
+    /// +-+-------+---+          +---------+---+
+    /// | |       |   |          |    2    | 1 |
+    /// |2|       | 1 |          +---------+---+
+    /// | |       |   |          |             |
+    /// +-+       +---+          +---------+---+
+    /// |4|       | 3 |          |    4    | 3 |
+    /// +-+-------+---+          +---------+---+
+    /// ```                      
+    ///
+    /// x is right and y is up, so the origin of the image is the lower left
+    /// corner of region 1.
+
+    pub fn new_regions(
+        xoff: i32,
+        yoff: i32,
+        size: i32,
+        pixels: &'a ImageBuffer<P, Vec<P::Subpixel>>,
+    ) -> [UpdateRegion<'a, P>; 4] {
+        // Check invariant
+        debug_assert!(0 <= xoff && xoff < size && 0 <= yoff && yoff <= size);
```

Comment:
nit: as discussed offline, I think the code could be simplified a little if we change the type of `xoff`, `yoff`, and `size` to `u32`.